make-based/run-all.sh

```diff
@@ -0,0 +1,193 @@
+    sleep 10
+
+    test_common_errors
+    if test $? -eq 1; then
+        return
+    fi
+
+    sudo kill -KILL $(pgrep -f "qemu-system.*$app") > /dev/null 2>&1
+
+    grep -E "^Hello World" < "$tmp_file" > /dev/null 2>&1 && echo "PASSED" || echo "FAILED"
+    rm -f "$tmp_file"
+}
+
+test_nginx()
+{
+    /bin/bash "do.sh" run < /dev/null > "$tmp_file" 2>&1 &
+    sleep 2
+
+    test_common_errors
+    if test $? -eq 1; then
+        return
+    fi
+
+    wget 172.44.0.2 > /dev/null 2>&1 && echo "PASSED" || echo "FAILED"
+    rm -f index.html 2> /dev/null
+    rm -f "$tmp_file"
+}
+
+test_httpreply()
+{
+    /bin/bash "do.sh" run < /dev/null > "$tmp_file" 2>&1 &
+    sleep 2
+
+    test_common_errors
+    if test $? -eq 1; then
+        return
+    fi
+
+    curl -s 172.44.0.2:8123 | grep "It works!" > /dev/null 2>&1 && echo "PASSED" || echo "FAILED"
+    rm -f "$tmp_file"
+}
+
+test_redis()
+{
+    /bin/bash "do.sh" run < /dev/null > "$tmp_file" 2>&1 &
+    sleep 2
+
+    test_common_errors
+    if test $? -eq 1; then
+        return
+    fi
+
+    pushd "$testing_dir" > /dev/null || exit 1
+    echo "PING" | timeout 10 ../utils/redis-cli -h 172.44.0.2 -p 6379 | grep "PONG" > /dev/null 2>&1 && echo "PASSED" || echo "FAILED"
+    popd > /dev/null || exit 1
+    rm -f "$tmp_file"
+}
+
+test_sqlite()
+{
+    # HACK: Clearly not the best way to do this.
+    # Find a way to capture the output of an application when ran using the
+    # `./do.sh` files.
+    (sleep 4; echo -e '.open chinook.db\nselect * from Album;\n.exit') | /bin/bash "do.sh" run > "$tmp_file" 2>&1 &
+
+    sleep 10
+    sudo kill -KILL $(pgrep -f "qemu-system.*$app") > /dev/null 2>&1
+
+    test_common_errors
+    if test $? -eq 1; then
+        return
+    fi
+
+    grep -E "^346\|Mozart: Chamber Music\|274" < "$tmp_file" > /dev/null 2>&1 && echo "PASSED" || echo "FAILED"
+    rm -f "$tmp_file"
+}
+
+test_micropython()
+{
+    # HACK: Clearly not the best way to do this.
+    # Find a way to capture the output of an application when ran using the
+    # `./do.sh` files.
+    (sleep 4; echo 'print("Hello World")') | /bin/bash "do.sh" run > "$tmp_file" 2>&1 &
+    sleep 10
+
+    test_common_errors
+    if test $? -eq 1; then
+        return
+    fi
+
+    sudo kill -KILL $(pgrep -f "qemu-system.*$app") > /dev/null 2>&1
+
+    grep -E "^Hello World" < "$tmp_file" > /dev/null 2>&1 && echo "PASSED" || echo "FAILED"
+    rm -f "$tmp_file"
+}
+
+
+test_automated()
+{
+    mkdir "$logging_dir" 2> /dev/null
+    while read -r script; do
+        app=$(echo "$script" | cut -d'-' -f2-)
+        pushd "$script" > /dev/null || exit 1
+
+        echo -n "Running $app... "
+        if test "$(type -t "test_$app")" = "function"; then
+            test_"$app"
+        else
+            test_booting
+        fi
+        sudo kill -KILL $(pgrep -f "qemu-system.*$app") > /dev/null 2>&1
+
+        popd > /dev/null || exit 1
+    done <<< "$(find . -type d -name "app*")"
```

Comment:
```suggestion
    done < <(find . -mindepth 1 -maxdepth 1 -type d)
```

Reply:
I'll also leave the `-name` here since otherwise it will try to run stuff from inside the `include/` and `logs/` directories too.